### PR TITLE
Add color commonality filter

### PR DESCRIPTION
**Pull request by russelldj**
This is one of several components of #1157 being broken up into smaller chunks. It implements filter which computes the frequency of similar pixels across the whole image.

### CODE REVIEW

File: arrows/vxl/color_commonality_filter.cxx
Comment (linus-sherrill) [inline arrows/vxl/color_commonality_filter.cxx:156]: Would pre-increment be better here?

File: arrows/vxl/color_commonality_filter.cxx
Comment (linus-sherrill) [inline arrows/vxl/color_commonality_filter.cxx:347]: It is not a big deal, but pre-increment would better match the style guide.

File: arrows/vxl/color_commonality_filter.h
Comment (linus-sherrill) [inline arrows/vxl/color_commonality_filter.h:23]: I think `algorithm_impl` is not needed here.

**Review by linus-sherrill — CHANGES_REQUESTED** [review]

**Review by mleotta — CHANGES_REQUESTED** [review]
Looks pretty good, but I've point out a few minor things to look at.

File: arrows/vxl/color_commonality_filter.cxx
Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx:446]: While there's nothing explicit in the style guide, I agree this looks weird. Also, the recommendations about boolean operator placement and comma placements between function arguments suggests this is incorrect.

File: arrows/vxl/color_commonality_filter.cxx
Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx:476]: I have not found any mention of this in the default configs so I'm comfortable removing it. If @mattdawkins has a use case for it I can add it back later.

Comment (mattdawkins) [reply arrows/vxl/color_commonality_filter.cxx]: I don't use it

**Comment by mleotta** [comment]
Jenkins test this please

**Review by mleotta — APPROVED** [review]

File: arrows/vxl/color_commonality_filter.cxx
Comment (mleotta) [inline arrows/vxl/color_commonality_filter.cxx:158]: It seems that this `decltype( +nj ) j{ 0 }` hack doesn't work to remove `const` for Visual Studio.  You could try:
`auto j = decltype(nj){0};`.

Personally, I think `decltype` also makes the code harder to read.  I don't see a strong argument for this over explicitly using `unsigned j = 0`, but I'm okay with `decltype` as long as it works reliably.

**Review by mleotta — CHANGES_REQUESTED** [review]
Use of `decltype( +const_var_name ){ 0 }` to get a non-const variable does not appear to work properly on the Windows dashboards.

File: arrows/vxl/color_commonality_filter.cxx
Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx:158]: @mwoehlke-kitware was advocating for it, so perhaps he has an argument for it.

However, I'm inclined to agree with you, especially in this context. Most times I've written `decltype` across these related PRs, it should evaluate to `unsigned`. From a maintenance perspective, I see the appeal if the argument to `decltype` is likely to be refactored, but in this case it's primarily `vil_image_views` which I doubt will change their API or be replaced with another type in the near future.

Comment (mleotta) [reply arrows/vxl/color_commonality_filter.cxx:158]: I know @mwoehlke-kitware really likes `auto` / `decltype`, and there are places where it really comes in handy.  However, I would argue that this is not really one of them.  `vil` has been doing loops over image pixels with `unsigned` type for 20+ years.  That type is not going to change and you could argue that `unsigned j=0` makes this loop better follow VXL style.  There are probably hundreds of examples in VXL of `for( unsigned j=0; j < nj; ++j, row += jstep )`.

**Comment by mleotta** [comment]
Jenkins test this please

**Review by mleotta — APPROVED** [review]

**Comment by mleotta** [comment]
I think this one is good to merge unless other reviewers would like to revisit it.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:233]: I don't love the chance of overflow here. It would be safer to upcast to 64-bit prior to the multiplication, and possibly to keep `value` as 64-bit until putting it back in `histogram[i]`.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:263]: I think this documentation either doesn't belong here, or shouldn't be doxygen. It could also stand to be rewrapped, and "occurring" is misspelled.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:289]: I'm not sure this comment is ideal, but at least...
```suggestion
    // Formulate gridded regions
```

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:320]: It might make more sense to instead put the remainder of the function in an `else`?

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:333]:
```suggestion
  bool use_external_hist = ( options.histogram != nullptr );
```
...but see next comment which would make this irrelevant.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:344]: The paranoiac in me suggests:
```suggestion
  std::vector< unsigned >* histogram = options.histogram;
  std::unique_ptr< std::vector< unsigned > > local_histogram;

  if( !histogram )
  {
    local_histogram.reset( new std::vector< unsigned >{ hist_size } );
    histogram = local_histogram.get();
  }
  else
  {
    histogram->resize( hist_size );
    std::fill( histogram->begin(), histogram->end(), 0 );
  }
```
:slightly_smiling_face: 
(...and lose the later explicit delete, of course.)

Regardless, resizing and filling the vector if we just `new`d it is redundant.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:358]: Not sure this is needed (or at least, doesn't need to be doxygen)

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:451]: Lines over length.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:521]: Would "unsupported" be better here? (Instead of "invalid".)

Same for the later exception.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:554]: This doesn't seem necessary, at least with any reasonable compiler. If it's really an issue, I would suggest instead shifting the complain-and-return code out of the `default:` and instead having a `default: break;`.

File: arrows/vxl/color_commonality_filter.h
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.h:33]: Missing periods.

**Review by mwoehlke-kitware — CHANGES_REQUESTED** [review]

**Comment by mwoehlke-kitware** [comment]
Although fairly far out of scope for this PR, I do also wonder about the "bright line cutoff" approach that this employs to "commonality". I wonder if it would be possible to implement this where the histogram bins are instead incremented in a weighted manner, i.e. a value right at the edge of one bin's range would instead increment that bin and the adjacent bin by roughly-equal values, while a value right in the middle of a bin would increment only that bin (keeping the total increment value constant, of course).

File: arrows/vxl/color_commonality_filter.cxx
Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx:498]: Ack, sorry. I thought I checked all the branches for this issue but apparently I missed this one.

**Comment by russelldj** [comment]
> Although fairly far out of scope for this PR, I do also wonder about the "bright line cutoff" approach that this employs to "commonality". I wonder if it would be possible to implement this where the histogram bins are instead incremented in a weighted manner, i.e. a value right at the edge of one bin's range would instead increment that bin and the adjacent bin by roughly-equal values, while a value right in the middle of a bin would increment only that bin (keeping the total increment value constant, of course).

I don't see an immediate benefit to this approach. My intuition could be wrong, but what you're suggesting seems like it's in effect creating new, shifted boundaries at non-integral locations.

**Comment by mwoehlke-kitware** [comment]
Say you have an int8 image that is 25% 127, 25% 128 and 50% 255. Realistically speaking, you should expect a uniform output; every pixel is similar to 50% of the other pixels. However, because 127 and 128 get binned separately, you would wind up with an image that is half 50% and half 25%.

On further consideration, I think my initial idea still has some issues, however I don't think the algorithm as-is is ideal either.

Again, though, I'm not proposing making any changes *now*. It's something to maybe think about for the indeterminate future.

File: arrows/vxl/color_commonality_filter.cxx
Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx]: I found the distribution a bit harder to read, but took advice otherwise.

File: arrows/vxl/color_commonality_filter.cxx
Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx]: I chose to keep the input specified as a value and not in bits, as the raw value is used frequently. However, I updated the documentation and error checking.

**Comment by mleotta** [comment]
Jenkins test this please

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [reply arrows/vxl/color_commonality_filter.cxx]: FWIW, I am not a fan of the 'commas on next line' style, and while obviously some people are¹, I'm not surprised the uncrustify configuration isn't optimized for this style.

Actually, `pos_constr_comma = trail` supposedly forbids this style, but it seems that option isn't actually working at present. (I tried various values, but none seemed to affect formatting.)

(¹ Many folks support this style even when argument commas use trailing placement. The argument is usually that it avoids a "spurious" change to the last initializer when adding a new member. IMHO its hiddeousness exceeds that small benefit to VCS history.)

Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx]: I did update it to member initialization here and have been trying to avoid this style across other developments.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:526]: Should this maybe do the same 'merge atop the default config' that `set_configuration` does?

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:260]: Uh... I think this is still broken for... actually, I'm not sure it works, ever, but I think it may just need:
```suggestion
    integer_log2( input_type_max ) + 1 -
```
Also, have you (at least mentally) verified this produces the correct result for `bitshift`? I haven't gone through the mental exercise of running this code in my head to verify that it isn't off-by-one; maybe you've done that already?

OTOH, I guess the tests are passing? I'd feel better knowing this has been micro-validated, though. If I'm wrong (always possible!), I'd at least feel better knowing someone is confident this is working.

File: arrows/vxl/color_commonality_filter.cxx
Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx:526]: I see the argument for that. However, I don't think I've seen this in any other filters so maybe it's worth just doing a batch PR to fix a number of them at once. 

In practice, my guess is check config is called after merging with the default, because I've never had key errors.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [reply arrows/vxl/color_commonality_filter.cxx]: > as the raw value is used frequently

Okay, so to "can we change the specification of resolution to be bits rather than values?", that'd be a "no" :slightly_smiling_face:. That's fine; was just a thought.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:280]: **Weak** suggestion:
```suggestion
      ( static_cast< unsigned long long >( histogram_scale_factor ) *
        static_cast< unsigned long long >( histogram[ i ] ) ) / sum;
```

File: arrows/vxl/color_commonality_filter.cxx
Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx:260]: I validated that this produces the correct bitshift, and the tests should fail violently if it's off. Since `integer_log2` truncates, and `input_type_max` will always be `2^n - 1`, the two suggested methods will behave the same as far as I can tell. I just thought this way was cleaner.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:395]: I think you missed:
```suggestion
      local_histogram.reset( new std::vector< unsigned >{ hist_size } );
      histogram = local_histogram.get();
```
...which is sort of important since it will SEGV otherwise :slightly_smiling_face:.

**Review by mwoehlke-kitware — CHANGES_REQUESTED** [review]
(-1 is just for the SEGV...)

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [reply arrows/vxl/color_commonality_filter.cxx:260]: Maybe I'm missing something, or maybe you're getting lucky with integer promotion (and not testing with 32- or 64-bit integers), but won't `input_type_max + 1` be `input_type_min` in at least some cases? Yes, it will be the same _if no overflow occurs_...

Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx:260]: Yup, you're right. I'll update it. 

The tests are only run on byte images, and AFAIK burnout has never really been pushed past that, so maybe that's something to consider in the future.

File: arrows/vxl/color_commonality_filter.cxx
Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx:395]: Ack, thanks. Batch adding suggestions was breaking.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:239]: (Minor, weak:)
```suggestion
  constexpr auto input_type_max = std::numeric_limits< InputType >::max();
  constexpr auto output_type_max = std::numeric_limits< OutputType >::max();
```

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:159]:
```suggestion
        step +=
          hist_steps[ p ] * static_cast< ptrdiff_t >( *plane >> bitshift );
```
(Feel free to wrap the other way if you prefer.)

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [reply arrows/vxl/color_commonality_filter.cxx:260]: Yeah, and I think you luck out because `char{255} + 1` produces `int{256}` due to integer promotion rules. I'm a little surprised actually there isn't a compiler warning when a constant expression overflows, but there are probably also instances of people doing that deliberately.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [inline arrows/vxl/color_commonality_filter.cxx:586]: BTW, is there a reason signed int{16,32,64} aren't supported? (I know why floating point aren't; the binning would be effectively impossible!)

Comment (russelldj) [reply arrows/vxl/color_commonality_filter.cxx:586]: Good catch, I have included them. FYI, there was initially code to do the calculations on floating points, but we removed it because it seemed unlikely to be used.

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [reply arrows/vxl/color_commonality_filter.cxx]: Actually, I think "in practice" it is only unit tests that actually call anyone's `set_configuration` :slightly_smiling_face:. And they probably don't check empty/default configurations.

I could see punting to a PR that fixes a bunch of these...

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [reply arrows/vxl/color_commonality_filter.cxx]: Okay, with the unit tests and the most recent update, I am reasonably confident this is working. Thanks!

File: arrows/vxl/color_commonality_filter.cxx
Comment (mwoehlke-kitware) [reply arrows/vxl/color_commonality_filter.cxx:591]: ...Did it look radically different?

I can't imagine how that would work without additional configuration. If you're just using floating point because you are weird, it seems you would need to specify in the configuration what the maximum value is, because it probably *isn't* `limits< double >::max()`. OTOH, if you're working with actual HDR stuff, you probably need some kind of logarithmic binning, because in HDR, 30.0 and 30.5 are "more similar" than 0.1 and 0.4.

**Review by mwoehlke-kitware — APPROVED** [review]

**Comment by dstoup** [comment]
I'm a bit confused by what happened here and sadly I can't see the CI any more which stinks. It looks like everything passed for here according to open.cdash, but this PR introduces a test for color_commonality_filter which has been failing on Windows since the merge. I will investigate the differences between the PR dashboard and the nightly to see why that was allowed to happen but either way, vxl:color_commonality_filter.color is broken and needs attention. @russelldj , can you take a look at this test to see why it fails and make a patch?

**Comment by dstoup** [comment]
Actually, @russelldj I will ping you offline. I want to make some CI script changes before you spend any time looking into the failure.

**Comment by dstoup** [comment]
Ok, I found the issue. We're filtering out 'Not Run' tests on the PR dashboard because all the Python tests that use the pre-GTest infrastructure are broken. For some reason the color_commonality_filter, which ends with SEGFAULT is getting filtered out as well. So, it looks like the PR falsely reported SUCCESS and the color_commonality_filter is indeed broken and in need of being fixed.

@russelldj  I will leave that fix up to you but let me know if you need any assistance. I will look into the CI script to see why we didn't fail when this PR was active.

**Comment by russelldj** [comment]
@dstoup Thanks for catching this. Is it still the case that this failure only happens on Windows? Do both the `color` and `gray` unit tests fail? I'm happy to take the discussion offline if that's better.

**Pull request merged**
